benchmark/test_biocypher_query_generation.py

```diff
@@ -13,142 +19,284 @@
 
 MODEL_NAMES = [
     "gpt-3.5-turbo",
-    "gpt-4",
+    # "gpt-4",
 ]
 
 
 @pytest.fixture(scope="module", params=MODEL_NAMES)
 def prompt_engine(request):
-    model_name = request.param
-    return BioCypherPromptEngine(
-        schema_config_or_info_path="test/test_schema_info.yaml",
-        model_name=model_name,
+    def setup_prompt_engine(kg_schema_path):
+        model_name = request.param
+        return BioCypherPromptEngine(
+            schema_config_or_info_path=kg_schema_path,
+            model_name=model_name,
+        )
+
+    return setup_prompt_engine
+
+
+def benchmark_already_executed(
+    task: str,
+    subtask: str,
+    model_name: str,
+    result_files: dict[str, pd.DataFrame],
+) -> bool:
+    """
+    Checks if the benchmark task and subtask for the model_name have already
+    been executed.
+
+    Args:
+        task (str): The benchmark task, e.g. "biocypher_query_generation"
+        subtask (str): The benchmark subtask, e.g. "entities"
+        model_name (str): The model name, e.g. "gpt-3.5-turbo"
+        result_files (dict[str, pd.DataFrame]): The result files
+
+    Returns:
+
+        bool: True if the benchmark task and subtask for the model_name has
+            already been run, False otherwise
+    """
+    task_results = result_files[f"benchmark/results/{task}.csv"]
+    task_results_subset = (task_results["model"] == model_name) & (
+        task_results["subtask"] == subtask
     )
+    return task_results_subset.any()
 
 
-def test_entity_selection(prompt_engine):
-    success = prompt_engine._select_entities(
-        question="Which genes are associated with mucoviscidosis?"
+def write_results_to_file(model_name: str, subtask: str, score: str):
+    with open(FILE_PATH, "a") as f:
+        f.write(f"{model_name},{subtask},{score}\n")
+
+
+def test_entity_selection(
+    prompt_engine, test_data_biocypher_query_generation, result_files
+):
+    subtask = inspect.currentframe().f_code.co_name
+
+    kg_schema_file_name = test_data_biocypher_query_generation[0]
+    prompt = test_data_biocypher_query_generation[1]
+    expected_entities = test_data_biocypher_query_generation[2]
+    prompt_engine = prompt_engine(
+        kg_schema_path=f"./benchmark/data/biocypher_query_generation/{kg_schema_file_name}"
     )
+
+    if benchmark_already_executed(
```

Comment:
We may need more expressive subtask names. The current implementation generates two tests for the entity selection, one for each of the two rows in the DF. However, both are called the same, `test_entity_selection`. This indicates that 1) the check here does not work correctly (otherwise it would skip the second run with the same subtask name), and 2) we need some way to generate a unique subtask name for each row.

Reply:
`test_end_to_end_query_generation_0` for the first row, `test_end_to_end_query_generation_1` for the second row, and so on. So now the check is working correctly and checking on the subtask test case level (row) and not only on the subtask level (file).